src/core/lib/resource_quota/memory_quota.cc

```diff
@@ -0,0 +1,375 @@
+// Reclaimer
+//
+
+ReclamationSweep::~ReclamationSweep() {
+  if (memory_quota_ != nullptr) {
+    memory_quota_->FinishReclamation(sweep_token_);
+  }
+}
+
+//
+// MemoryRequest
+//
+
+namespace {
+size_t RoundUp(size_t size, size_t block_size) {
+  return (size + block_size - 1) / block_size * block_size;
+}
+size_t RoundDown(size_t size, size_t block_size) {
+  return size / block_size * block_size;
+}
+}  // namespace
+
+MemoryRequest MemoryRequest::WithBlockSize(size_t block_size) const {
+  MemoryRequest r(*this);
+  r.block_size_ = block_size;
+  return r;
+}
+
+MemoryRequest MemoryRequest::Increase(size_t amount) const {
+  MemoryRequest r(min_ + amount, max_ + amount);
+  r.block_size_ = block_size_;
+  return r;
+}
+
+//
+// ReclaimerQueue
+//
+
+const ReclaimerQueue::Index ReclaimerQueue::kInvalidIndex;
+
+ReclaimerQueue::Index ReclaimerQueue::Insert(
+    RefCountedPtr<MemoryAllocator> allocator, ReclamationFunction reclaimer) {
+  MutexLock lock(&mu_);
+  Index index;
+  if (free_entries_.empty()) {
+    index = entries_.size();
+    entries_.emplace_back(std::move(allocator), std::move(reclaimer));
+  } else {
+    index = free_entries_.back();
+    free_entries_.pop_back();
+    Entry& entry = entries_[index];
+    entry.allocator = std::move(allocator);
+    entry.reclaimer = std::move(reclaimer);
+  }
+  if (queue_.empty()) waker_.Wakeup();
+  queue_.push(index);
+  return index;
+}
+
+ReclamationFunction ReclaimerQueue::Cancel(Index index,
+                                           MemoryAllocator* allocator) {
+  MutexLock lock(&mu_);
+  if (index >= entries_.size()) return nullptr;
+  Entry& entry = entries_[index];
+  if (entry.allocator.get() != allocator) return {};
+  entry.allocator.reset();
+  return std::move(entry.reclaimer);
+}
+
+Poll<ReclamationFunction> ReclaimerQueue::PollNext() {
+  MutexLock lock(&mu_);
+  while (true) {
+    if (queue_.empty()) {
+      waker_ = Activity::current()->MakeNonOwningWaker();
+      return Pending{};
+    }
+    Index index = queue_.front();
+    queue_.pop();
+    free_entries_.push_back(index);
+    Entry& entry = entries_[index];
+    if (entry.allocator != nullptr) {
+      entry.allocator.reset();
+      return std::move(entry.reclaimer);
+    }
+  }
+}
+
+//
+// MemoryAllocator
+//
+
+MemoryAllocator::MemoryAllocator(RefCountedPtr<MemoryQuota> memory_quota)
+    : memory_quota_(memory_quota) {
+  Reserve(sizeof(MemoryQuota));
+}
+
+MemoryAllocator::~MemoryAllocator() {
+  Release(sizeof(MemoryQuota));
+  GPR_ASSERT(free_bytes_.load(std::memory_order_acquire) == taken_bytes_);
+  memory_quota_->Return(taken_bytes_);
+}
+
+void MemoryAllocator::Orphan() {
+  {
+    absl::MutexLock lock(&memory_quota_mu_);
+    for (int i = 0; i < kNumReclamationPasses; i++) {
+      memory_quota_->reclaimers_[i].Cancel(reclamation_indices_[i], this);
+    }
+  }
+  InternallyRefCounted<MemoryAllocator>::Unref();
+}
+
+size_t MemoryAllocator::Reserve(MemoryRequest request) {
+  while (true) {
```

Comment:
For my education: While this busy loop forever on persistent failures, or is it gonna block on the reclaimer future below?

Reply:
Reservation is infallible, so failures won't stop this... but races might (we replenish, then some other allocator takes it) - I expect the concurrent allocation rate against an allocator won't be crazy though, so a few iterations ought to resolve it.

Reply:
I see. Sounds fine. We can later adjust if needed.